matrix_sdk/src/client.rs

```diff
@@ -371,6 +370,74 @@ impl<'a> SyncSettings<'a> {
     }
 }
 
+/// Configuration for requests the `Client` makes.
+///
+/// This sets how often and for how long a request should be repeated. As well as how long a
+/// successful request is allowed to take.
+///
+/// By default requests are retried indefinitely and use no timeout.
+///
+/// # Example
+///
+/// ```
+/// # use matrix_sdk::RequestConfig;
+/// # use std::time::Duration;
+/// // This sets makes requests fail after a single send request and sets the timeout to 30s
+/// let request_config = RequestConfig::new()
+///     .disable_retry()
+///     .timeout(Duration::from_secs(30));
+/// ```
+#[derive(Copy, Clone, Default)]
+pub struct RequestConfig {
+    pub(crate) timeout: Option<Duration>,
+    pub(crate) retry_limit: Option<u64>,
+    pub(crate) retry_timeout: Option<Duration>,
+}
+
+#[cfg(not(tarpaulin_include))]
+impl Debug for RequestConfig {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut res = fmt.debug_struct("RequestConfig");
+
+        res.field("timeout", &self.timeout)
+            .field("retry_limit", &self.retry_limit)
+            .field("retry_timeout", &self.retry_timeout)
+            .finish()
+    }
+}
+
+impl RequestConfig {
+    /// Create a new default `RequestConfig`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// This is a convince method to disable the retries of a request. Setting the `retry_limit` to `0`
+    /// has the same effect.
+    pub fn disable_retry(mut self) -> Self {
+        self.retry_limit = Some(0);
+        self
+    }
+
+    /// The number of times a request should be retried. The default is no limit
+    pub fn retry_limit(mut self, retry_limit: u64) -> Self {
+        self.retry_limit = Some(retry_limit);
+        self
+    }
+
+    /// Set a timeout duration for all HTTP requests. The default is no timeout.
+    pub fn timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = Some(timeout);
+        self
+    }
+
+    /// Set a timeout for how long a request should be retried. The default is no timeout.
```

Comment:
We should clarify what the timeout here means.
```suggestion
    /// Set a timeout for how long a request should be retried. The default is no timeout, meaning requests are retried forever.
```